### PR TITLE
[MM-31987] Allow camera use for jitsi

### PR DESCRIPTION
**Pull request by Willyfrog**
**Summary**

Enable entitlements for using the camera and microphone for third party plugins like jitsi

**Issue link**

[MM-31987](https://mattermost.atlassian.net/browse/MM-31987)

**Test Cases**

- go to a community server and join a jitsi call. 
- first time it should request access to camera and mic.

**Additional Notes**

@esethna please review messages for requesting access to camera and mic, I'm sure there is a better wording for it.

Also, I've added  a v4.6 milestone, but I'd wait until our meeting to discuss when and how to deploy this one.

### CODE REVIEW

File: electron-builder.json [hunk 1]
```diff
@@ -77,7 +77,13 @@
       }
     ],
     "hardenedRuntime": true,
-    "gatekeeperAssess": true
+    "gatekeeperAssess": true,
+    "entitlements": "./build/entitlements.mac.plist",
+    "entitlementsInherit": "./build/entitlements.mac.plist",
+    "extendInfo": {
+      "NSMicrophoneUsageDescription": "Third party plugins (like jitsi) will use the microphone to provide in-app videoconference",
+      "NSCameraUsageDescription": "Third party plugins (like jitsi) will use the microphone to provide in-app videoconference"
```
Comment (larkox) [inline electron-builder.json:85]: Potentially, plugins can use camera and microphone for many things, not just videoconferencing. We might want to give a more vague approach to this. 0/5

File: build/entitlements.mac.plist [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>com.apple.security.device.microphone</key>
+	<true/>
+	<key>com.apple.security.device.camera</key>
+	<true/>
+	<key>com.apple.security.device.audio-input</key>
```
Comment (larkox) [inline build/entitlements.mac.plist:9]: Just curiosity. What covers audio-input and why it is needed? Isn't audio input the same as microphone?

**Review by larkox — APPROVED** [review]
LGTM, just a consideration in the wording and one question purely out of curiosity.

File: build/entitlements.mac.plist (hunk 2 above)
Comment (Willyfrog) [reply build/entitlements.mac.plist:9]: I wish I could tell, from my tests I wasn't being able to share the camera until I added this one. It's description is that it is used to record audio from the mic, so apparently it is redundant but it didn't work without it 🤷‍♂️. My guess is that jitsi allows to change its source (when you have more than one), but it is just a guess.

File: electron-builder.json (hunk 1 above)
Comment (Willyfrog) [reply electron-builder.json:85]: Definitely there can be a better wording, I'll wait for @esethna comment on it.

**Review by devinbinnie — APPROVED** [review]

File: electron-builder.json (hunk 1 above)
Comment (esethna) [reply electron-builder.json:85]: @Willyfrog do you have a screenshot of how this prompt would appear?

File: electron-builder.json [hunk 3]
```diff
@@ -77,7 +77,13 @@
       }
     ],
     "hardenedRuntime": true,
-    "gatekeeperAssess": true
+    "gatekeeperAssess": true,
+    "entitlements": "./build/entitlements.mac.plist",
+    "entitlementsInherit": "./build/entitlements.mac.plist",
+    "extendInfo": {
+      "NSMicrophoneUsageDescription": "Third party plugins (like jitsi) will use the microphone to provide in-app videoconference",
```
Comment (esethna) [inline electron-builder.json:84]: Microphone access may be used by Mattermost plugins, such as Jitsi video conferencing.

File: electron-builder.json (hunk 1 above)
Comment (esethna) [inline electron-builder.json:85]: Camera access may be used by Mattermost plugins, such as Jitsi video conferencing.

**Review by esethna — CHANGES_REQUESTED** [review]

File: electron-builder.json (hunk 1 above)
Comment (Willyfrog) [reply electron-builder.json]: I don't! I didn't take one when it asked and now it doesn't allow me to get it to ask again.

It is a standard OS modal, I don't really have access to it. I can share a build with you so you can test if you prefer

**Review by esethna — APPROVED** [review]

**Review by jgilliam17 — APPROVED** [review]
Thanks @Willyfrog 
Tested, looks good to merge.
- Verified allow camera and microphone access for Jitsi. Attaching images for visibility. 
Screen Shot 2021-01-18 at 1 56 49 PM
Screen Shot 2021-01-18 at 1 56 30 PM

**Comment by esethna** [comment]
@amyblais we should ship a v4.6 dot release for this. @Willyfrog please let Amy know if there's any other PRs to take in the dot release

**Comment by amyblais** [comment]
There's this PR, but Guillermo may have more details: https://github.com/mattermost/desktop/pull/1439.

**Comment by Willyfrog** [comment]
> There's this PR, but Guillermo may have more details: #1439.

yes, those two are the ones to be shipped in a new dot release.

**Comment by amyblais** [comment]
@Willyfrog This was accidentally cherry-picked to ``browserview`` branch instead of ``release-4.6``.

**Comment by Willyfrog** [comment]
> @Willyfrog This was accidentally cherry-picked to ``browserview`` branch instead of ``release-4.6``.

Thanks for noticing, but I have it cherry picked into my local release-4.6 branch and it is waiting for the second PR to be added as well. I could push it if you like, but that will trigger a build and make some messages in the release channel. Let me know if you prefer to have it pushed already.

**Comment by amyblais** [comment]
No, thanks for the clarification. 👍 It can be cherry picked when the other PR is ready.

**Pull request merged**